src/xdocs-examples/resources-noncompilable/com/puppycrawl/tools/checkstyle/checks/design/sealedshouldhavepermitslist/Example1.java

```diff
@@ -0,0 +1,27 @@
+/*xml
+<module name="Checker">
+  <module name="TreeWalker">
+    <module name="SealedShouldHavePermitsList"/>
+  </module>
+</module>
+*/
+
+
+//non-compiled with javac: Compilable with Java17
+package com.puppycrawl.tools.checkstyle.checks.design.sealedshouldhavepermitslist;
+
+// xdoc section -- start
+public class Example1 {
+  sealed class A {} // violation
+  final class B extends A {}
+  final class C extends A {}
+  final class D { } // this can extend A, so as any other class in the same file
+}
+
+class CorrectedExample1 {
+  sealed class A permits B, C {} // ok, explicitly declared permitted subclasses
+  final class B extends A {}
+  final class C extends A {}
+  final class D { } // this can not extend A
+}
```

Comment:
please add example:
```
public class Example2 {
  sealed class A {} // violation
  // bunch of code
  final class D { } // this can extend A, so as any other class in the same file
  // bunch of code
}
```
and this becomes very intereting .... we will annoy users to have any sealed class. 
and force  users to have 
```
public class Example1 {
  sealed class A {} permits SealingA
  final class SealingA extends A {}
  final class D { } // this can extend A, so as any other class in the same file
}
```

where users need to make fake SealingA as just double confirmation that nobody else will extend.
this is not ok.

What you tried to do is catch only cases like:
```
public class Example1 {
  sealed class A {}  
  final class LazyExtensionA extends A {} // violation , 'allowance by permit clause at A should be explicit'
}
```
and
```
public class Example2 {
  // no violation, as nobody even try to use hack:
  //        ` it implicitly allows any class within the same compilation unit to extend it.` 
  sealed class A {} 
  // bunch of code
  final class D { } // this can extend A, and until it is not extending, there is no problem
  // bunch of code
}
```

https://docs.oracle.com/javase/specs/jls/se21/html/jls-8.html#jls-8.1.6

>If C is not an enum class, then its permitted direct subclasses are those classes declared in the same compilation unit as C ([§7.3](https://docs.oracle.com/javase/specs/jls/se21/html/jls-7.html#jls-7.3)) which have a canonical name ([§6.7](https://docs.oracle.com/javase/specs/jls/se21/html/jls-6.html#jls-6.7)) and whose direct superclass is C.
That is, the permitted direct subclasses are inferred as the classes in the same compilation unit that specify C as their direct superclass. The requirement for a canonical name means that no local classes or anonymous classes will be considered.


So we are fighting with implicit nuances of jdk, and by means of Check we want to do all explicit.
But Explicit declaration of permits is required when we detecting extending, so code owner just got lazy to explicitly declare permits.

Reply:
I got confused a little because we have a non-compile able example here

> please add example:
> 
> public class Example2 {
>   sealed class A {} // violation
>   // bunch of code
>   final class D { } // this can extend A, so as any other class in the same file
>   // bunch of code
> }
> 

---

```
D:\CS\test cat src/Test.java
public class Test {
    sealed class A { }
    final class D { }
}
D:\CS\test javac src/Test.java
src\Test.java:2: error: sealed class must have subclasses
    sealed class A { }
           ^
1 error
```
To make the example compilable it should be:

```
D:\CS\test cat src/Test.java
public class Test {
    sealed class A { }   // violation by us, it should be ( sealed class A permits D { } ) 
    final class D extends A { }
}
```
btw, the above example is a violation by IDEA inspections.

Does this satisfy your thoughts? If not then please share a compilable example with what you expect vs what we actually do

Reply:
@romani please prove that your concerns are valid by using `javac` on your examples above.

Reply:
I think his thought are around idea that we may have a sealed class that have no subclasses in this case to satisfy the check we will make a fake class and add it to the permitted list of the sealed class 
but this will not compile sealed class mush have subclasses
@romani please confirm if I understand you correctly

Reply:
Yes. Current logic kind forbid usage of single sealed class. It demands subclass.

Reply:
Then, please finish the review and merge :)

Reply:
@romani This is also non-compilable
```
public class Example2 {
  sealed class A {} // violation
  // bunch of code
  
}
```
---

```
D:\CS\test cat src/Test.java
public class Test {
    sealed class A {}
}
D:\CS\test javac src/Test.java
src\Test.java:2: error: sealed class must have subclasses
    sealed class A {}
           ^
1 error
```

> We need tests on sealed classes at top level of same file (compilation unit).
https://github.com/checkstyle/checkstyle/blob/108ec4cd543c93d44f2edec9fbea13070337441d/src/test/resources-noncompilable/com/puppycrawl/tools/checkstyle/checks/design/sealedshouldhavepermitslist/InputSealedShouldHavePermitsListTopLevelSealedClass.java

Reply:
very smart compiler, awesome.